services/rollout-service/pkg/datadogdora/datadogdora.go

```diff
@@ -0,0 +1,205 @@
+	"golang.org/x/sync/errgroup"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+type Config struct {
+	URL         string
+	APIKey      string
+	Concurrency int
+	MaxEventAge time.Duration
+}
+
+func New(config Config) *Subscriber {
+	sub := &Subscriber{
+		url:           "",
+		ready:         nil,
+		state:         nil,
+		maxAge:        0,
+		now:           nil,
+		group:         errgroup.Group{},
+		apiKey:        "",
+		doraAPI:       nil,
+		RepositoryUrl: "",
+	}
+	sub.group.SetLimit(config.Concurrency)
+	sub.url = config.URL
+	sub.apiKey = config.APIKey
+	sub.ready = func() {}
+	sub.maxAge = config.MaxEventAge
+	sub.now = time.Now
+	configuration := datadog.NewConfiguration()
+	configuration.SetUnstableOperationEnabled("v2.CreateDORADeployment", true)
+	apiClient := datadog.NewAPIClient(configuration)
+	sub.doraAPI = datadogV2.NewDORAMetricsApi(apiClient)
+
+	return sub
+}
+
+type Subscriber struct {
+	group         errgroup.Group
+	apiKey        string
+	url           string
+	RepositoryUrl string
+	// The ready function is needed to sync tests
+	ready func()
+	state map[service.Key]*service.BroadcastEvent
+	// The maximum age of events that should be considered. If 0,
+	// all events are considered.
+	maxAge time.Duration
+	// Used to simulate the current time in tests
+	now func() time.Time
+	// used to report dora metrics
+	doraAPI *datadogV2.DORAMetricsApi
+}
+
+func (s *Subscriber) Subscribe(ctx context.Context, b *service.Broadcast) error {
+	if s.state == nil {
+		s.state = map[service.Key]*service.BroadcastEvent{}
+	}
+	for {
+		err := s.subscribeOnce(ctx, b)
+		select {
+		case <-ctx.Done():
+			return err
+		default:
+		}
+	}
+}
+
+func (s *Subscriber) subscribeOnce(ctx context.Context, b *service.Broadcast) error {
+	event, ch, unsub := b.Start()
+	defer unsub()
+	for _, ev := range event {
+		if ev.IsProduction != nil && *ev.IsProduction {
+			s.state[ev.Key] = ev
+		}
+	}
+	s.ready()
+	for {
+		select {
+		case <-ctx.Done():
+			return s.group.Wait()
+		case ev, ok := <-ch:
+			if !ok {
+				return s.group.Wait()
+			}
+			if ev.IsProduction == nil || !*ev.IsProduction {
+				continue
+			}
+			if s.maxAge != 0 &&
+				ev.ArgocdVersion != nil &&
+				ev.ArgocdVersion.DeployedAt.Add(s.maxAge).Before(s.now()) {
+				continue
+			}
+			if shouldNotify(s.state[ev.Key], ev) {
+				s.group.Go(s.notify(ctx, ev))
+			}
+			s.state[ev.Key] = ev
+		}
+	}
+}
+
+func shouldNotify(old *service.BroadcastEvent, nu *service.BroadcastEvent) bool {
+	// check for fields that must be present to generate the request
+	if nu.ArgocdVersion == nil || nu.IsProduction == nil || nu.ArgocdVersion.SourceCommitId == "" {
+		return false
+	}
+	if old == nil || old.ArgocdVersion == nil || old.IsProduction == nil {
+		return true
+	}
+	if old.ArgocdVersion.SourceCommitId != nu.ArgocdVersion.SourceCommitId || old.ArgocdVersion.DeployedAt != nu.ArgocdVersion.DeployedAt {
+		return true
+	}
+	return false
```

Comment:
```suggestion
	if old == nil || old.ArgocdVersion == nil || old.IsProduction == nil || old.ArgocdVersion.SourceCommitId != nu.ArgocdVersion.SourceCommitId || old.ArgocdVersion.DeployedAt != nu.ArgocdVersion.DeployedAt {
		return true
	}
	return false
```

Reply:
The first if sounds and looks redundant to me, we can simply omit it

Reply:
I think there is beneift having this explict even though you can have it on one line. 
Anyway I will adapt